pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -17,106 +17,176 @@ limitations under the License.
-	"k8s.io/cloud-provider-gcp/pkg/util/gkenetworkparamsetutil"
+	"k8s.io/cloud-provider-gcp/crd/client/network/clientset/versioned/typed/network/v1alpha1"
 	"k8s.io/cloud-provider-gcp/providers/gce"
+
 	controllersmetrics "k8s.io/component-base/metrics/prometheus/controllers"
 )
 
 // Controller manages GKENetworkParamSet status.
 type Controller struct {
-	gkeNetworkParamsInformer networkinformers.GKENetworkParamSetInformer
+	gkeNetworkParamsInformer cache.SharedIndexInformer
 	networkClientset         networkclientset.Interface
-	informerStartFn          func(stopCh <-chan struct{})
 	gceCloud                 *gce.Cloud
+	queue                    workqueue.RateLimitingInterface
 }
 
 // NewGKENetworkParamSetController returns a new
 func NewGKENetworkParamSetController(
 	networkClientset networkclientset.Interface,
-	gkeNetworkParamsInformer networkinformers.GKENetworkParamSetInformer,
-	informerStartFn func(stopCh <-chan struct{}),
+	gkeNetworkParamsInformer cache.SharedIndexInformer,
 	gceCloud *gce.Cloud,
 ) *Controller {
 
 	return &Controller{
 		networkClientset:         networkClientset,
 		gkeNetworkParamsInformer: gkeNetworkParamsInformer,
-		informerStartFn:          informerStartFn,
 		gceCloud:                 gceCloud,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "gkenetworkparamset"),
 	}
 
 }
 
 // Run starts an asynchronous loop that monitors and updates GKENetworkParamSet in the cluster.
-func (c *Controller) Run(stopCh <-chan struct{}, controllerManagerMetrics *controllersmetrics.ControllerManagerMetrics) {
+func (c *Controller) Run(numWorkers int, stopCh <-chan struct{}, controllerManagerMetrics *controllersmetrics.ControllerManagerMetrics) {
 	defer utilruntime.HandleCrash()
 
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+	defer c.queue.ShutDown()
+
 	klog.Infof("Starting gkenetworkparamset controller")
 	defer klog.Infof("Shutting down gkenetworkparamset controller")
 	controllerManagerMetrics.ControllerStarted("gkenetworkparamset")
 	defer controllerManagerMetrics.ControllerStopped("gkenetworkparamset")
 
-	c.gkeNetworkParamsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    gkenetworkparamsetutil.CreateAddGKENetworkParamSetHandler(c.onGKENetworkParamSetAdded),
-		UpdateFunc: gkenetworkparamsetutil.CreateUpdateGKENetworkParamSetHandler(c.onGKENetworkParamSetUpdated),
+	c.gkeNetworkParamsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
 	})
 
-	c.informerStartFn(stopCh)
+	if !cache.WaitForNamedCacheSync("gkenetworkparamset", stopCh, c.gkeNetworkParamsInformer.HasSynced) {
```

Comment:
`c.gkeNetworkParamsInformer.HasSynced` use to be added to the struct so can be mocked on the tests , using `func () {return true}` , ref https://github.com/kubernetes/kubernetes/blob/6ce055d62d8608842db18901c948c6982c74a164/pkg/controller/endpoint/endpoints_controller.go#L136-L141

Reply:
is it possible that the CI is stuck because it never initializes